src/net/http/UploadFile.cpp

```diff
@@ -0,0 +1,243 @@
+  as published by the Free Software Foundation; either version 2
+  of the License, or (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+}
+*/
+
+#include "UploadFile.hpp"
+#include "Request.hxx"
+#include "Handler.hxx"
+#include "Operation/Operation.hpp"
+#include "io/FileOutputStream.hxx"
+#include "thread/Mutex.hxx"
+#include "thread/Cond.hxx"
+#include "util/NumberParser.hpp"
+#include "util/ScopeExit.hxx"
+#include "util/StaticString.hxx"
+
+#include "LogFile.hpp"
+
+#include <cassert>
+#ifndef MAX_PATH  // #include <limits.t>?
+# define MAX_PATH  0x200
+#endif
+
+class UploadFileHandler final : public CurlResponseHandler {
+  uint8_t* buffer;
+  const size_t max_size;
+  
+  size_t posted = 0;
+  size_t post_count = 0;
+
+  size_t received = 0;
+
+  OperationEnvironment &env;
+
+  Mutex mutex;
+  Cond cond;
+
+  std::exception_ptr error;
+
+  bool done = false;
+  
+ public:
+  UploadFileHandler(void* _buffer, size_t _max_size,
+    OperationEnvironment& _env)
+    :buffer((uint8_t*)_buffer), max_size(_max_size), env(_env) {}
+
+  size_t GetReceived() const {
+    return received;
+  }
+
+  void Cancel() noexcept {
+    const std::lock_guard<Mutex> lock(mutex);
+    done = true;
+    cond.notify_one();
+  }
+
+  void Wait() {
+    std::unique_lock<Mutex> lock(mutex);
+    cond.wait(lock, [this]{ return done; });
+
+    if (error)
+      std::rethrow_exception(error);
+  }
+
+  /* virtual methods from class CurlResponseHandler */
+  void OnHeaders(unsigned status,
+                 std::multimap<std::string, std::string> &&headers) override {
+    if (auto i = headers.find("content-length"); i != headers.end())
+      env.SetProgressRange(ParseUint64(i->second.c_str()));
+  }
+
+  void OnData(ConstBuffer<void> data) override {
+    size_t remaining = max_size - received;
+    if (remaining == 0 || done)
+      throw Pause{};
+
+    if (data.size > remaining)
+      data.size = remaining;
+
+    memcpy(buffer + received, data.data, data.size);
+    received += data.size;
+
+    env.SetProgressRange(received);
+  }
+
+  void OnEnd() override {
+    const std::lock_guard<Mutex> lock(mutex);
+    done = true;
+    cond.notify_one();
+  }
+
+  void OnError(std::exception_ptr e) noexcept override {
+    const std::lock_guard<Mutex> lock(mutex);
+    error = std::move(e);
+    try {
+      if (error) {
+        std::rethrow_exception(error);
+      }
+    }
+    catch (const std::exception& e) {
+         std::snprintf(reinterpret_cast<char*>(buffer),
+           max_size, "%s", e.what());
+    }
```

Comment:
....to get the exception info (e.what) from exception_ptr, following the instruction at https://en.cppreference.com/w/cpp/error/exception_ptr
I need this info for a correct interpretion in case the curl error.

Reply:
But ... why? Why rethrow, and why copy to that buffer? Nothing of this appears to make sense.

Reply:
... and what if this isn't a `std::exception`?

Reply:
This happens in Request.cxx in function Done(...)
```
void
CurlRequest::Done(CURLcode result) noexcept
{
	Stop();

	try {
		if (result != CURLE_OK) {
			StripRight(error_buffer);
			const char *msg = error_buffer;
			if (*msg == 0)
				msg = curl_easy_strerror(result);
			throw FormatRuntimeError("CURL failed: %s", msg);
		}

		FinishBody();
	} catch (...) {
		state = State::CLOSED;
		handler.OnError(std::current_exception());
	}
}
```
The handler.OnError is called with the std::exception_ptr := std::current_exception() 
Maybe no 'error = std::move(e);' is necessary, I can call the rethrow-... directly. But up to now I didn't find any other method to read the error_message from curl with an untouched CurlRequest::Done() function. 
If you have a better solution for this, you are very welcome - I only followed the guide...
From my point of view there are only std::exceptions possible, because I didn't find another caller of OnError...
And the message buffer in this case is empty, so I filled it with this error information - with CURLE_OK this is filled with the (json) response of the server in this case

Reply:
To convert an arbitrary exception to a string, you can use this one:
https://github.com/XCSoar/XCSoar/blob/53affb6cdbb88715a217587c87557ef36546a03c/src/util/Exception.hxx#L125-L132
... but that still doesn't answer "why copy to that buffer?"
> And the message buffer in this case is empty, so I filled it with this error information

This isn't an explanation. This doesn't make sense.

Reply:
Thanks, this 'GetFullMessage'  works! But of course you are right: I used a general exception handler (catch(...)) in the caller routine and want to have this information... 
Now I will use a more specified exception handler (catch(const std::exception&)) - and can analyse this exception code directly - so I don't need this copy to the buffer anymore :)

Reply:
Why copy to that buffer? That's a bad way of handling errors.